README.md
# ![logo][] Logfella

[logo]: https://raw.githubusercontent.com/dustinmoris/Logfella/master/assets/logfella-64x64.png

[![NuGet Badge](http://buildstats.info/nuget/logfella)](https://www.nuget.org/packages/Logfella/)

[![Build History](https://buildstats.info/github/chart/dustinmoris/Logfella?branch=master)](https://github.com/dustinmoris/Logfella/actions?query=branch%3Amaster)

Logfella is a small logging library specifically designed to work well with [Google Cloud Logging](https://cloud.google.com/logging/).

If you are running a .NET Core or ASP.NET Core application inside GCP and want to write structured logs into [Google Cloud Logging](https://cloud.google.com/logging/) then this library will offer you a more complete integration point than .NET Core's out of the box `ILogger` providers.

Logfella writes structured logs directly to stdout in a specific JSON format which GCP's log analyzer can natively understand and process accordingly. Logfella offers support for the full range of [Google's severity levels](#severity-levels), automatic propagation of exceptions into [Google Error Reporting](https://cloud.google.com/error-reporting), full integration with `Microsoft.Extensions.Logging` libraries and much more.

#### NOTE:

This library is an early version mostly built for personal use and neither the docs nor the APIs have been fully tested or fine tuned for a more global exposure.

That being said, it has been written with flexibility and extensibility 100% in mind and hopefully will be able to suit more than just my immediate and personal use case. It's been open sourced under Apache 2.0 and feedback and/or PRs are more than welcome.

## Benefits of Logfella

- Write structured logs into [Google Cloud Logging](https://cloud.google.com/logging/) when hosted in GCP
- Automatic error propagation into [Google Error Reporting](https://cloud.google.com/error-reporting) when exceptions are being logged
- Support of custom log labels for advanced log filtering
- Support of the full range of severity levels of GCP (9 in total)
- Automatically sets the correct timestamp and service context information for all Google Cloud Logs
- Seamlessly integrates with .NET Core's logging providers (`ILogger` and `ILoggerProvider`)
- Optimised for C# and F# applications so that logging feels natural from an object-oriented as well as from a functional architecture
- `ILogWriter` can be injected either via DI or called directly from the static `Log` instance
- 100% customisable and extensible
- Uses .NET Core's new `System.Text.Json` library for faster JSON serialisation
- Supports log aggregation via an optional correlation ID which can be set on a `LogWriter` instance. The additional `RequestScopedLogWriterMiddleware` makes it extremely easy to compute (or inherit) a correlation ID for all logs of a given HTTP request pipeline
- HTTP request data can be logged with each log entry as part of the ASP.NET Core request pipeline

## Overview

In order to start logging with Logfella one has to create a logger object of type `ILogWriter`. By default this library comes with four loggers out of the box:

- `GoogleCloudLogWriter`
- `ConsoleLogWriter`
- `MultiLogWriter`
- `NullLogWriter`

The `GoogleCloudLogWriter` is the main logger which should be used when an application is running inside the Google Cloud Platform. The `ConsoleLogWriter` is a simple console writer which can be swapped for the `GoogleCloudLogWriter` during development. The `MultiLogWriter` is a convenience class to configure multiple log writers simultaneously. Lastly the `NullLogWriter` is a void implementation which is used to prevent a `NullReferenceException` from occuring when no other log writer has been configured.

## GoogleCloudLogWriter

The `GoogleCloudLogWriter` is a relatively unexciting console logger which outputs log entries in a certain JSON format into `stdout` which Google Cloud's log analyzer knows how to parse and interpret accordingly.

The benefit of writing logs in GCP in this way is that the process of logging itself has almost no impact on latency, doesn't require any additional network calls, doesn't require any obscure dependencies or convoluted infrastructure configuration and it doesn't require log batching like most other log providers do. Logs will get ingested into the Google Cloud Logging console almost in real-time and allow the StackDriver analyzer do its job best.

The easiest way to create a new `GoogleCloudLogWriter` is to use the `Create(...)` builder method:

###### C#

```csharp
var logWriter = GoogleCloudLogWriter.Create(Severity.Info);
```

###### F#

```fsharp
let logWriter = GoogleCloudLogWriter.Create Severity.Info
```

This will create a basic GCP log writer with the minimum set of features. The severity parameter supports the full range of GCP's severity levels (see below).

### Severity Levels

Google Cloud Logging supports the following severity levels:

| Severity       | Value | Description |
| :------------- | :---- | :---------- |
| `Default` | 0 | The log entry has no assigned severity level. |
| `Debug` | 100 | Debug or trace information. |
| `Info` | 200 | Routine information, such as ongoing status or performance. |
| `Notice` | 300 | Normal but significant events, such as start up, shut down, or a configuration change. |
| `Warning` | 400 | Warning events might cause problems. |
| `Error` | 500 | Error events are likely to cause problems. |
| `Critical` | 600 | Critical events cause more severe problems or outages. |
| `Alert` | 700 | A person must take an action immediately. |
| `Emergency` | 800 | One or more systems are unusable. |

### Configuring the GoogleCloudLogWriter

The `GoogleCloudLogWriter` offers several builder methods which can be used to add additional features to the logger instance. The `GoogleCloudLogWriter` class itself is immutable and each builder method invocation will create a new instance which makes it safe to use in a multi threaded environment (which is useful when a new log writer has to be configured per HTTP request - more on this topic in the [ASP.NET Core section](#using-with-aspnet-core)).

#### Error Reporting Feature

When logging an `Exception` object with the `Error` severity level or above (`Error`, `Critical`, `Alert` and `Emergency`) then the error can get automatically propagated into [Google Error Reporting](https://cloud.google.com/error-reporting). This is a handy feature if a team wants to take additional advantage of error tracking and automatic error notifications in the Google Cloud Platform.

In order to enable this feature the `GoogleCloudLogWriter` has to be configured with a service context which the Error Reporting APIs rely on. This is done via the `AddServiceContext("serviceName", "serviceVersion")` builder method:

###### C#

```csharp
var logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Info)
        .AddServiceContext(
            "<your-service-name>",
            "<your-service-version>");
```

###### F#

```fsharp
let logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Info)
        .AddServiceContext(
            "<your-service-name>",
            "<your-service-version>")
```

The service name and service version are two properties solely used by the Google Error Reporting APIs and will not appear in the structured logs otherwise. For more information please check out the [Google Cloud Error Reporting formatting docs](https://cloud.google.com/error-reporting/docs/formatting-error-messages).

#### Labels

Labels are a neat feature to "tag" log entries with useful meta data which can be used to sieve through a large number of logs. Common use cases would be to label (= tag) each log entry with the name of the application/service and the version of it. This can be extremely handy if a team is analysing logs and trying to attribute certain log entries to a specific application version (e.g. trying to understand which version introduced a newly occuring error).

The `AddLabels(...)` builder method can be used to set a whole `IDictionary<string, string>` of labels on the `GoogleCloudLogWriter` instance, otherwise the `AddLabel(key, value)` builder method can be used to add an additional label to the existing dictionary of labels:

###### C#

```csharp
var logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Info)
        .AddServiceContext(
            "<your-service-name>",
            "<your-service-version>")
        .AddLabels(
            new Dictionary<string, string>
                {
                    {"appName", "MyApp1234"},
                    {"appVersion", "1.0.0"}
                });

// Add an extra label later:
var newLogWriter = logWriter.AddLabel("foo", "bar");
```

###### F#

```fsharp
let logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Debug)
        .AddServiceContext(appName, appVersion)
        .AddLabels(
            dict [
                "appName", "MyApp1234"
                "appVersion", "1.0.0"
            ])

// Add an extra label later:
let logWriter' = logWriter.AddLabel("foo", "bar")
```

#### Timestamp Settings

By default the `GoogleCloudLogWriter` will set an ISO 8601 formatted UTC timestamp for each log entry, however if omitted then the Google Cloud log analyzer will set its own timestamp in RFC 3339 UTC "Zulu" format - accurate to nanoseconds and by 3 nanosecond digits more accurate than what .NET Core can achieve natively.

Computing the timestamp in .NET Core will be closer to the actual time when the log entry is being written, whereas letting Google Cloud set its own timestamp will enable a higher accuracy between different log entries. It's a trade-off and if the latter is preferred then the `UseGoogleCloudTimestamp()` builder method can be used to configure the log writer accordingly:

###### C#

```csharp
var logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Info)
        .AddServiceContext(
            "<your-service-name>",
            "<your-service-version>")
        .AddLabels(
            new Dictionary<string, string>
                {
                    {"appName", "MyApp1234"},
                    {"appVersion", "1.0.0"}
                })
        .UseGoogleCloudTimestamp();
```

###### F#

```fsharp
let logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Debug)
        .AddServiceContext(appName, appVersion)
        .AddLabels(
            dict [
                "appName", "MyApp1234"
                "appVersion", "1.0.0"
            ])
        .UseGoogleCloudTimestamp()
```

#### Correlation ID

Sometimes it can be very helpful to tag log entries with a specific ID in order to group correlated log entries together. This can be done by adding the correlation ID feature to the `GoogleCloudLogWriter`:

###### C#

```csharp
var logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Info)
        .AddServiceContext(
            "<your-service-name>",
            "<your-service-version>")
        .AddLabels(
            new Dictionary<string, string>
            {
                {"appName", "MyApp1234"},
                {"appVersion", "1.0.0"}
            })
        .UseGoogleCloudTimestamp()
        .AddCorrelationId("<some id>");
```

###### F#

```fsharp
let logWriter =
    GoogleCloudLogWriter
        .Create(Severity.Debug)
        .AddServiceContext(appName, appVersion)
        .AddLabels(
            dict [
                "appName", "MyApp1234"
                "appVersion", "1.0.0"
            ])
        .UseGoogleCloudTimestamp()
        .AddCorrelationId("<some id>")
```

The correlation ID will be set as part of the data dictionary of each log entry, which can be queried by the `jsonPayload.data.correlationId` property in the Google Cloud Console. If the name `correlationId` collides with another data piece with the same name, or if one simply wants to specify a different name, then an additional string argument can be passed into the `AddCorrelationId(value, key)` builder method to set a custom key:

```
logWriter.AddCorrelationId("<some id>", "requestId")
```

This feature is most useful in combination with the `RequestScopedLogWriterMiddleware` (see [ASP.NET Core](#using-with-aspnet-core)).

#### HTTP Context

Log entries can also include information about the HTTP request if they've been written during the ASP.NET Core request pipeline. In order to do so pass the `HttpContext` object into the `AddHttpContext(ctx)` builder method:

```
logWriter.AddHttpContext(context);
```

This feature makes most sense as part of the `RequestScopedLogWriterMiddleware` (see [ASP.NET Core](#using-with-aspnet-core)).

## Using with ASP.NET Core

Configuring Logfella as the preferred logging library in a .NET Core application is as easy as registering the library via the `UseLogfella()` extension method of an `IHostBuilder` or alternatively of an `ILoggingBuilder` object:

###### C#

```csharp
Host.CreateDefaultBuilder(args)
    .UseLogfella()
    .ConfigureWebHost(
        webBuilder =>
        {
            webBuilder
                .UseKestrel()
                .UseStartup<Startup>();
        })
    .Build()
    .Run();
```

###### F#

```fsharp
Host.CreateDefaultBuilder()
    .UseLogfella()
    .ConfigureWebHost(
        fun webHostBuilder ->
            webHostBuilder
                .UseKestrel()
                .Configure(configureApp)
                .ConfigureServices(configureServices)
                |> ignore)
    .Build()
    .Run()
```

Additionally the `RequestScopedLogWriterMiddleware` can be registered early in the HTTP pipeline to decorate all log entries with HTTP context information and an optional correlation ID.

For a more detailed example of how to use Logfella as part of an ASP.NET core web application please refer to the MVC example in C# or the Giraffe application in F#, which are both found in the `/samples` directory of this repository.

## Examples

This is an example of what logs can be like when using Logfella:

![screenshot-1](https://raw.githubusercontent.com/dustinmoris/Logfella/master/screenshots/logfella-demo-1.png)

![screenshot-2](https://raw.githubusercontent.com/dustinmoris/Logfella/master/screenshots/logfella-demo-2.png)

## License

 [Apache 2.0](https://github.com/dustinmoris/Logfella/master/LICENSE)


RELEASE_NOTES.md
# 7.1.0

- Upgraded to .NET 5

Logfella `7.1.0` and `7.0.0` are identical in functionality, but `7.1.0` is targetting .NET 5 only now.

# 7.0.0

- Renamed `RequestBasedLogWriterMiddleware` to `RequestScopedLogWriterMiddleware`
- Changed the `RequestLoggingMiddleware` to use an object of `RequestLoggingOptions` for its configuration.

The extension method `app.UseRequestLogging(...)` has changed its signature from multiple optional parameters to accepting a single `Action<RequestLoggingOptions>` method to configure the middleware.

# 6.2.0

The request logging middleware accepts two new parameters to filter HTTP headers from the logs:

- `includeHttpHeaders`
- `excludeHttpHeaders`

Example usage:

```
app.UseRequestLogging(
        excludeHttpHeaders: HashSet.New("Authorization", "Cookie", "X-ApiKey"))
    .UseRouting()
    .UseEndpoints(
        endpoints => { endpoints.MapControllers(); });
```

The `HashSet.New` helper method is just a nicer way of initialising a `HashSet<string>` of http headers in the example above. It's particularly helpful in F# projects where creating a new hash set is not as straight forward as `new HashSet<string> { "..." }` due to the awkward interop between F# and C# type classes.

# 6.1.0

Always creating a safe copy of the data dictionary to prevent a `NotSupportedException` when adding additional items in F# where the `dict` type only returns an immutable version of `IDictionary<TKey, TValue>`

# 6.0.1

Fixed `UseRequestLogging` extension method.

# 6.0.0

- Naming is hard, therefore renamed `LogWriterPerRequestMiddleware` to `RequestBasedLogWriterMiddleware`
- Extended the `RequestLoggingMiddleware` to have an option for logging before and after other middleware has been invoked (helpful to debug X-Forwarded-For settings and middleware)
- Added a new `MuteLogWriter` and the `Mute` builder class to filter/mute log entries based on a predicate (experimental)

# 5.0.1

Fixed private `ctor` of `GoogleCloudLogWriter`

# 5.0.0

- Renamed `IncludeHttpRequest` to `AddHttpContext`
- Renamed `Log.SetGlobalLogWriter` to `Log.SetDefaultLogWriter`
- Added `AsLogWriter` convenience method for F# to cast any log writer to the common interface
- Made all log writers immutable
- Added immutable builder methods to the `GoogleCloudLogWriter` to allow a more dynamic, flexible and safe composition API
- Fixed double `/` bug in the request path of the `httpRequest` object
- Improved error logging

# 4.0.0

- Removed duplicate error logging
- Setting the `@type` property for exception logs
- Added ability to log request data when using the per-request logging middleware
- Minor breaking changes through re-ordering and renaming some method parameters

# 3.0.0

- Changed the `SetLogWriter` and `GetLogWriter` methods to a single `LogWriter` property with setter and getter.
- Added an internal `AsyncLocal<ILogWriter>` instance for per-request log writer instantiation (can be used to set a correlation ID per HTTP request). Use the new `PerRequestLogWriterMiddleware` to set up a new `ILogWriter` instance for each request pipeline.
- The `UseLogfella` extension method will automatically register a transient `ILogWriter` dependency.
- Each log writer accepts an optional `correlationIdKey` and `correlationId` parameter in their constructors.
- The coloring of console logs can be disabled for the `ConsoleLogWriter` now.
- Sealed all `ILogWriter` implementations.

# 2.0.0

Extremely minor, but breaking API change of making `IHostBuilder` and `ILoggingBuilder` extension methods return a non `void` type to support configuration chaining.

# 1.0.0

First version. Hello Logfella!

src/Logfella/ILogWriter.cs
using System;
using System.Collections.Generic;

namespace Logfella
{
    /// <summary>
    /// Structured Log Writer.
    /// </summary>
    public interface ILogWriter
    {
        /// <summary>
        /// Logs a message and a set of given data and/or an exception with the given severity.
        /// </summary>
        void Log(Severity severity, string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (0) The log entry has no assigned severity level.
        /// </summary>
        void Default(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (0) The log entry has no assigned severity level.
        /// </summary>
        void Default(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (0) The log entry has no assigned severity level.
        /// </summary>
        void Default(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (100) Debug or trace information.
        /// </summary>
        void Debug(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (100) Debug or trace information.
        /// </summary>
        void Debug(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (100) Debug or trace information.
        /// </summary>
        void Debug(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (200) Routine information, such as ongoing status or performance.
        /// </summary>
        void Info(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (200) Routine information, such as ongoing status or performance.
        /// </summary>
        void Info(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (200) Routine information, such as ongoing status or performance.
        /// </summary>
        void Info(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (300) Normal but significant events, such as start up, shut down, or a configuration change.
        /// </summary>
        void Notice(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (300) Normal but significant events, such as start up, shut down, or a configuration change.
        /// </summary>
        void Notice(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (300) Normal but significant events, such as start up, shut down, or a configuration change.
        /// </summary>
        void Notice(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (400) Warning events might cause problems.
        /// </summary>
        void Warning(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (400) Warning events might cause problems.
        /// </summary>
        void Warning(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (400) Warning events might cause problems.
        /// </summary>
        void Warning(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (500) Error events are likely to cause problems.
        /// </summary>
        void Error(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (500) Error events are likely to cause problems.
        /// </summary>
        void Error(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (500) Error events are likely to cause problems.
        /// </summary>
        void Error(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (600) Critical events cause more severe problems or outages.
        /// </summary>
        void Critical(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (600) Critical events cause more severe problems or outages.
        /// </summary>
        void Critical(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (600) Critical events cause more severe problems or outages.
        /// </summary>
        void Critical(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (700) A person must take an action immediately.
        /// </summary>
        void Alert(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (700) A person must take an action immediately.
        /// </summary>
        void Alert(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (700) A person must take an action immediately.
        /// </summary>
        void Alert(string message, IDictionary<string, object> data, Exception ex = null);

        /// <summary>
        /// (800) One or more systems are unusable.
        /// </summary>
        void Emergency(string message, params Tuple<string, object>[] data);

        /// <summary>
        /// (800) One or more systems are unusable.
        /// </summary>
        void Emergency(string message, Exception ex, params Tuple<string, object>[] data);

        /// <summary>
        /// (800) One or more systems are unusable.
        /// </summary>
        void Emergency(string message, IDictionary<string, object> data, Exception ex = null);
    }
}